### PR TITLE
CI: Test for the health of all code base URLs

### LINKED ISSUE #2126: CI: Proactively check the code base for dead URLs
As suggested by #2121, we ought to proactively maintain healthy URLs.

### PR DESCRIPTION
**Pull request by fallwith**
- Add a new unit test class for testing the health of all URLs defined in the code base. See the comments in `test/new_relic/healthy_urls_test.rb` for more details
- Update `ci_cron.yml` to set a `CI_CRON` env var to true
- Update `.ruboycop.yml` to permit `throw` and `catch` calls without parens
- Add new `skip_unless_ci_cron` helper to permit tests to only run via `ci_cron.yml`
- Add new `skip_unless_newest_ruby` helper to permit tests to only run on the newest Ruby
- Add new `agent_root` helper to permit tests to quickly get to the agent root
- Fix 5 URLs that the new test caught

resolves #2126

### CODE REVIEW

**Comment by fallwith** [comment]
The new test will only run via CI cron, and only on the newest Ruby version available.

Using a branch with a single bad URL, here is a demonstration: https://github.com/newrelic/newrelic-ruby-agent/actions/runs/5516187929

**Review by kaylareopelle — COMMENTED** [review]
I really like this test and the new helpers you've added. Approved as-is. Comments are more out of curiosity rather than a need for anything to change.

**Review by kaylareopelle — APPROVED** [review]

File: lib/new_relic/agent/instrumentation/memcache.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@
 # each with slightly different APIs and semantics.
 # See:
 #     http://www.deveiate.org/code/Ruby-MemCache/ (Gem: Ruby-MemCache)
-#     http://seattlerb.rubyforge.org/memcache-client/ (Gem: memcache-client)
+#     https://github.com/mperham/memcache-client (Gem: memcache-client)
```
[the remark this answers is not in the document]
Comment (fallwith) [reply lib/new_relic/agent/instrumentation/memcache.rb:9]: Yeah, I'd be interested in knowing more about the history there. I think that repo is a fork of another fork of the original RubyForge code (which Mike also has commit access to). For whatever reason, it's what's linked to from the Rubygems.org page's "Homepage" link, so I'm using it here.

File: test/new_relic/healthy_urls_test.rb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+# URL to be completed.
+#
+# Enable DEBUG for additional verbosity
+
+require_relative '../test_helper'
+
+class HealthyUrlsTest < Minitest::Test
+  ROOT = File.expand_path('../../..', __FILE__).freeze
+  FILENAMES = %w[
+    baselines
```
[the remark this answers is not in the document]
Comment (fallwith) [reply test/new_relic/healthy_urls_test.rb:28]: It's at `test/performance/script/baselines`. I haven't checked, but possibly it's used to power the perf tests' setting of a baseline from which a subsequent set of results is compared to.

File: test/new_relic/healthy_urls_test.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    Dockerfile
+    Envfile
+    Gemfile
+    Guardfile
+    install_mysql55
+    LICENSE
+    mega-runner
+    newrelic
+    newrelic_cmd
+    nrdebug
```
[the remark this answers is not in the document]
Comment (fallwith) [reply test/new_relic/healthy_urls_test.rb:40]: Yeah, these ones are really just Ruby scripts so they're eligible for having URLs in the source code. For now I went with an allowlist approach of filenames and extensions for specifying what to scan for URLs. In the future we may want to reverse this and have a blocklist of things like image files or perhaps even take a quick peek at the file's encoding and make a judgment call from that.

File: test/new_relic/healthy_urls_test.rb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  EXTENSIONS = %w[
+    css
+    erb
+    gemspec
+    haml
+    html
+    js
+    json
+    md
+    proto
```
[the remark this answers is not in the document]
Comment (fallwith) [reply test/new_relic/healthy_urls_test.rb:55]: Yes! Not only can we, but we have, and with a URL in fact! There is currently only 1 file at `infinite_tracing/lib/proto/infinite_tracing.proto`. That file has a URL in the source code comments at the top.

**Comment by olleolleolle** [comment]
Cool that you built this.

File: test/new_relic/healthy_urls_test.rb [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  end
+
+  def real_url?(url)
+    return false if url.match?(IGNORED_URL_PATTERN)
+
+    true
+  end
+
+  def gather_urls
+    Dir.glob(File.join(ROOT, '**', '*')).each_with_object({}) do |file, urls|
```
Comment (olleolleolle) [inline test/new_relic/healthy_urls_test.rb:104]: Minor: Can default the `urls` keys to be an empty array, ready to be << onto!

Example from docs: https://rubyapi.org/3.2/o/hash
`h = Hash.new { |hash, key| "Default value for #{key}" }`

Change we can do:
`each_with_object(Hash.new { |h, k| h[k] = [] })`.

Which then would mean that `urls[url] ||= []` could be omitted in  line 155

Comment (fallwith) [reply test/new_relic/healthy_urls_test.rb:104]: Thanks, @olleolleolle! Implemented with #2132.

Using `Hash.new([])` in conjunction with `+= []` from [Ulysee's blog post](https://medium.com/klaxit-techblog/a-headache-in-ruby-hash-default-values-bf2706660392) looks pretty slick too, but the Ruby docs example approach is more readable and predictable in its behavior, so we'll go with it.

```ruby
LETTERS = ('a'..'z').to_a

hash1 = LETTERS.each_with_object(Hash.new { |h, k| h[k] = [] }) { |l, h| h[l] << true }
hash2 = LETTERS.each_with_object(Hash.new([])) { |l, h| h[l] += [true] }
hash1 == hash2
```

**Pull request merged**